p4-16/psa/PSA.mdk

```diff
@@ -2576,3 +2576,127 @@ Certainly if there are long durations of time when egress decides to
+packet every clock cycle.  The latency might be hundreds of clock
+cycles from starting a packet until it is complete.  P4 tables in such
+ASICs are typically implemented using logic such as TCAMs and SRAMs.
+TCAM designs can do 1 search per clock cycle.  The lowest area and
+power SRAMs can do 1 read or 1 write per clock cycle.
+
+While there are "multi ported" SRAM designs that can be read and/or
+written multiple times per clock cycle, these have a noticeable
+increase in area and power over the "singled ported" designs that are
+limited to 1 access per cycle.  If multi ported TCAM designs even
+exist, they are likely even a higher cost premium of single ported
+TCAMs than for SRAMs.  They typical way to achieve higher search rates
+is via parallelism, by creating multiple copies of the desired TCAM,
+which is a linear increase in area and power (at least).
+
+Due to these issues, if one wishes to create a switch ASIC that
+achieves a packet processing rate of N times the clock rate, e.g. 2
+billion packets per second in a 1 GHz ASIC, the most straightforward
+way to do this is to take advantage of the fact that packet processing
+is (mostly) an embarrasingly parallel
+problem[^EmbarrassinglyParallel], and create a device with N modules,
+each modules processing packets at 1 billion packets per second.
+
+[^EmbarrassinglyParallel]: https://en.wikipedia.org/wiki/Embarrassingly_parallel
+
+A PSA device designed in this way would typically have N ingress
+pipeline modules, plus N egress pipeline modules.  It is common to
+assign multiple physical ports of the switch ASIC to each pipeline in
+a hard-wired fashion, e.g. a device with 32 100 Gigabit Ethernet ports
+might physically hard-wire ports 0 through 15 to pipeline 0, and ports
+16 through 31 to pipeline 1.  All packets received on ports 0 through
+15 will be processed by ingress pipeline 0, then go to the packet
+buffer (if they were not dropped in ingress), then go to egress
+pipeline 0 if ingress chose output port 0 through 15, or go to egress
+pipeline 1 if ingress chose output port 16 through 31.
+
+In such a device, typically you will want the same P4 program to be
+run on every one of these N ingress pipelines, and every one of the N
+egress pipelines.
+
+It is physically possible in such a device for the control plane to
+install different table entries in the different pipelines, and there
+are use cases where this can help achieve higher scale in number of
+usable table entries.  For example, perhaps you have an ingress table
+with the ingress port as one of the fields in its search key.  If this
+is the case, the packet processing behavior is the same even if table
+entries for port X are installed only in the ingress pipeline that
+processes packets from port X.  Installing that table entry in the
+other pipelines would be an unnecessary use of table space, since it
+could never be matched.  Whether the software of such a PSA device
```

Comment:
What is the software here? The control plane or the data plane program?

Reply:
Changed "software" to "device-dependent control plane software" in hopes of clarifying this.